source/brailleDisplayDrivers/nlseReaderZoomax.py

```diff
@@ -0,0 +1,295 @@
+	LOC_JOYSTICK_KEYS: 1,
+	LOC_DEVICE_ID: 16,
+	LOC_SERIAL_NUMBER: 8,
+}
+
+KEY_NAMES = {
+	LOC_ROUTING_KEYS: None,
+	LOC_ROUTING_KEY: None,
+	LOC_DISPLAY_KEYS: ("d1", "d2", "d3", "d4", "d5", "d6"),
+	LOC_BRAILLE_KEYS: (
+		"bl",
+		"br",
+		"bs",
+		None,
+		"s1",
+		"s2",
+		"s3",
+		"s4",  # byte 1
+		"b1",
+		"b2",
+		"b3",
+		"b4",
+		"b5",
+		"b6",
+		"b7",
+		"b8",
+	),  # byte 2
+	LOC_JOYSTICK_KEYS: ("up", "left", "down", "right", "select"),
+}
+
+
+class BrailleDisplayDriver(braille.BrailleDisplayDriver):
+	_dev: hwIo.IoBase
+	name = "nlseReaderZoomax"
+	# Translators: Names of braille displays.
+	description = _("NLS eReader Zoomax")
+	isThreadSafe = True
+	supportsAutomaticDetection = True
+
+	@classmethod
+	def registerAutomaticDetection(cls, driverRegistrar: bdDetect.DriverRegistrar):
+		driverRegistrar.addUsbDevices(
+			bdDetect.DeviceType.SERIAL,
+			{
+				"VID_1A86&PID_7523",  # CH340
+			},
+		)
+
+		driverRegistrar.addBluetoothDevices(lambda m: m.id.startswith("NLS eReader Z"))
+
+	@classmethod
+	def getManualPorts(cls):
+		return braille.getSerialPorts()
+
+	def _connect(self, port):
+		for portType, portId, port, portInfo in self._getTryPorts(port):
+			try:
+				self._dev = hwIo.Serial(
+					port,
+					baudrate=BAUD_RATE,
+					bytesize=serial.EIGHTBITS,
+					parity=serial.PARITY_NONE,
+					stopbits=serial.STOPBITS_ONE,
+					timeout=TIMEOUT,
+					writeTimeout=TIMEOUT,
+					onReceive=self._onReceive,
+				)
+			except EnvironmentError:
+				log.info("Port not yet available.")
+				log.debugWarning("", exc_info=True)
+				if self._dev:
+					self._dev.close()
+				continue
+
+			self._sendRequest(LOC_PROTOCOL_ONOFF, False)
+			self._sendRequest(LOC_PROTOCOL_ONOFF, True)
+			self._sendRequest(LOC_PROTOCOL_ONOFF, True)
+			self._sendRequest(LOC_REPEAT_ALL)
+
+			for i in range(5):
+				self._dev.waitForRead(TIMEOUT)
+				if self.numCells:
+					break
+
+			if self.numCells:
+				log.info(
+					"Device connected via {type} ({port})".format(
+						type=portType,
+						port=port,
+					),
+				)
+				return True
+			log.info("Device arrival timeout")
+			self._dev.close()
+		return False
+
+	def __init__(self, port="auto"):
+		log.info("nlseReaderZoomax Init")
+		super(BrailleDisplayDriver, self).__init__()
+		self.numCells = 0
+		self._deviceID: Optional[str] = None
+		self._dev = None
+
+		for i in range(CONNECT_RETRIES):
+			if self._connect(port):
+				break
+			else:
+				time.sleep(TIMEOUT_BETWEEN_RETRIES)
+
+		self._keysDown = {}
+		self._ignoreKeyReleases = False
+
+	def terminate(self):
+		try:
+			super(BrailleDisplayDriver, self).terminate()
+			try:
+				self._sendRequest(LOC_PROTOCOL_ONOFF, False)
+			except EnvironmentError:
+				pass
+		finally:
+			self._dev.close()
+
+	def _sendRequest(self, command: bytes, arg: Union[bytes, bool, int] = b""):
+		"""
+		:type command: bytes
+		:type arg: bytes | bool | int
+		"""
+		typeErrorString = "Expected param '{}' to be of type '{}', got '{}'"
```

Comment:
Could you change this to use named rather than positional fields?
```suggestion
		TYPE_ERROR_TEMPLATE = "Expected parameter '{paramName}' to be of type '{expectedType}', got '{actualType}'"
```